src/base/Datatypes.ml

```diff
@@ -232,7 +232,7 @@ module SnarkTypes = struct
   let scilla_g1point_to_ocaml g1p =
     match g1p with
     | ADTValue ("Pair", [ pxt; pyt ], [ ByStrX px; ByStrX py ])
-      when type_equiv pxt scalar_type && type_equiv pyt scalar_type
+      when type_assignable pxt scalar_type && type_assignable pyt scalar_type
```

Comment:
Are these functions (i.e., the associated builtins) not accessible for developers? I'm ok with with using `type_equiv`, but the question is whether someone would want to use an Address (viewed as a ByStr20) as a point in this calculation.

In essence it boils down to whether we should make an exception to the rule that we always al allow implicit coercion from Address to ByStr20. As I mentioned in the PR description I think we should remove that implicit coercion in the next major version, but we can't do that without introducing builtins for expliti type coercion, and that would not be backward compatible.

Reply:
>the question is whether someone would want to use an Address (viewed as a ByStr20) as a point in this calculation.

No. The `ByStrX` here, even when it's `ByStr20` will not / cannot refer to addresses. It is actually just a byte string for snark computations.

Regarding allowing implicit coercion from `Address` to `ByStr20`, I'm not convinced that it shouldn't be allowed. It looks to be fully compatible with the idea that you can move (assign) from a type with more information (say `Address with x : Uint32, y : Uint32` end) to a type with less information (`Address with x : Uint32 end`). The basic `ByStr20` just loses all extra information about the address. I can't see how it could be a reason for the programmer to introduce bugs.

Reply:
> > the question is whether someone would want to use an Address (viewed as a ByStr20) as a point in this calculation.
> 
> No. The `ByStrX` here, even when it's `ByStr20` will not / cannot refer to addresses. It is actually just a byte string for snark computations.

Got it. Then it should indeed be `type_equiv`.

> 
> Regarding allowing implicit coercion from `Address` to `ByStr20`, I'm not convinced that it shouldn't be allowed. It looks to be fully compatible with the idea that you can move (assign) from a type with more information (say `Address with x : Uint32, y : Uint32` end) to a type with less information (`Address with x : Uint32 end`). The basic `ByStr20` just loses all extra information about the address. I can't see how it could be a reason for the programmer to introduce bugs.

Except that there is a difference between `ByStr20 with end` and `ByStr20`. The former is an address on the network which has (dynamically) been verified as containing a contract (though we don't know anything about its fields). The latter is just any byte string which may or may not refer to an address on the network, and if it does, then it may refer to any address, including user addresses or unused addresses.

The point I am trying to make is that with this change, addresses and byte strings are conceptually different, and should therefore be treated as such, and since we don't allow implicit type coercions anywhere else in Scilla, I don't think we should allow them here, either.

Reply:
> The point I am trying to make is that with this change, addresses and byte strings are conceptually different, and should therefore be treated as such, and since we don't allow implicit type coercions anywhere else in Scilla, I don't think we should allow them here, either.

I agree :+1: 

>The latter is just any byte string which may or may not refer to an address on the network, and if it does, then it may refer to any address, including user addresses or unused addresses.
Let us not differentiate contract and non-contract addresses. That's because we may want to fetch `_balance` of non-contract addresses.